src/base/ClpMap.h

```diff
@@ -0,0 +1,260 @@
+
+    using IndexItem = std::pair<Key, EntriesIterator>;
+    /// key:entry_position mapping for fast entry lookups by key
+    using Index = std::unordered_map<Key, EntriesIterator, std::hash<Key>, std::equal_to<Key>, PoolingAllocator<IndexItem> >;
+    using IndexIterator = typename Index::iterator;
+
+    static uint64_t MemoryCountedFor(const Key &, const Value &);
+
+    void trim(uint64_t wantSpace);
+    void erase(const IndexIterator &);
+    IndexIterator find(const Key &);
+
+    /// cached entries, including expired ones, in LRU order
+    Entries entries_;
+
+    /// entries_ positions indexed by the entry key
+    Index index_;
+
+    /// entry TTL to use if none provided to add()
+    Ttl defaultTtl_ = std::numeric_limits<Ttl>::max();
+
+    /// the maximum memory we are allowed to use for all cached entries
+    uint64_t memLimit_ = 0;
+
+    /// the total amount of memory we currently use for all cached entries
+    uint64_t memUsed_ = 0;
+};
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+ClpMap<Key, Value, MemoryUsedBy>::ClpMap(const uint64_t capacity, const Ttl defaultTtl):
+    defaultTtl_(defaultTtl)
+{
+    assert(defaultTtl >= 0);
+    setMemLimit(capacity);
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+void
+ClpMap<Key, Value, MemoryUsedBy>::setMemLimit(const uint64_t newLimit)
+{
+    assert(newLimit >= 0);
+    if (memUsed_ > newLimit)
+        trim(memLimit_ - newLimit);
+    memLimit_ = newLimit;
+}
+
+/// \returns the index position of an entry identified by its key (or end())
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+typename ClpMap<Key, Value, MemoryUsedBy>::IndexIterator
+ClpMap<Key, Value, MemoryUsedBy>::find(const Key &key)
+{
+    const auto i = index_.find(key);
+    if (i == index_.end())
+        return i;
+
+    const auto entryPosition = i->second;
+    if (!entryPosition->expired()) {
+        if (entryPosition != entries_.begin())
+            entries_.splice(entries_.begin(), entries_, entryPosition);
+        return i;
+    }
+    // else fall through to cleanup
+
+    erase(i);
+    return index_.end();
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+const Value *
+ClpMap<Key, Value, MemoryUsedBy>::get(const Key &key)
+{
+    const auto i = find(key);
+    if (i != index_.end()) {
+        const auto &entry = *(i->second);
+        return &entry.value;
+    }
+    return nullptr;
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+uint64_t
+ClpMap<Key, Value, MemoryUsedBy>::MemoryCountedFor(const Key &k, const Value &v)
+{
+    // approximate calculation (e.g., containers store wrappers not value_types)
+    const auto storageSz = sizeof(typename Entries::value_type) + k.length() + MemoryUsedBy(v);
+    const auto indexSz = sizeof(typename Index::value_type) + k.length();
+    return storageSz + indexSz;
```

Comment:
@yadij, the above counts the key length twice. The current code nearly assumes SBuf use (by not optimizing key copies and by calling length()). When Key is SBuf, there are no copies, so we should count the key length _once_. Should i change this to count the key once? Or byte the bullet and outsource key size calculation to the map creator?

This is not a minor concern because certificate validator keys, for example, are often 5-50KB in size (because they contain PEM certificates)!

Reply:
I am okay with reducing to one key.length(). But would place a code comment in with that change to indicate why we are able to ignore one of the two keys.

Reply:
> I am okay with reducing to one key.length().

Done at d9ea1e5.